src/daft-csv/src/local.rs

```diff
@@ -0,0 +1,924 @@
+//     \n character and go back to 2.
+// 2c. If we hit the end of a chunk without coming to a decision on whether we found a valid CSV record,
+//     we simply move on to the next chunk of bytes and try to find a valid CSV record there. This is a
+//     simplification that makes the implementation a lot easier to maintain.
+
+// Default size for CSV buffers.
+const DEFAULT_CSV_BUFFER_SIZE: usize = SLABSIZE; // 4MiB. Like SLABSIZE, this can be tuned.
+
+/// Reads a single local CSV file in a non-streaming fashion.
+pub async fn read_csv_local(
+    uri: &str,
+    convert_options: Option<CsvConvertOptions>,
+    parse_options: CsvParseOptions,
+    read_options: Option<CsvReadOptions>,
+    io_client: Arc<IOClient>,
+    io_stats: Option<IOStatsRef>,
+    max_chunks_in_flight: Option<usize>,
+) -> DaftResult<Table> {
+    let stream = stream_csv_local(
+        uri,
+        convert_options.clone(),
+        parse_options.clone(),
+        read_options,
+        io_client.clone(),
+        io_stats.clone(),
+        max_chunks_in_flight,
+    )
+    .await?;
+    let tables = Box::pin(stream);
+    // Apply limit.
+    let limit = convert_options.as_ref().and_then(|opts| opts.limit);
+    let mut remaining_rows = limit.map(|limit| limit as i64);
+    use futures::TryStreamExt;
+    let collected_tables = tables
+        .try_take_while(|result| {
+            match (result, remaining_rows) {
+                // Limit has been met, early-terminate.
+                (_, Some(rows_left)) if rows_left <= 0 => futures::future::ready(Ok(false)),
+                // Limit has not yet been met, update remaining limit slack and continue.
+                (table, Some(rows_left)) => {
+                    remaining_rows = Some(rows_left - table.len() as i64);
+                    futures::future::ready(Ok(true))
+                }
+                // (1) No limit, never early-terminate.
+                // (2) Encountered error, propagate error to try_collect to allow it to short-circuit.
+                (_, None) => futures::future::ready(Ok(true)),
+            }
+        })
+        .try_collect::<Vec<_>>()
+        .await?
+        .into_iter()
+        .collect::<Vec<_>>();
+    // Handle empty table case.
+    if collected_tables.is_empty() {
+        let (schema, _, _) = get_schema_and_estimators(
+            uri,
+            &convert_options.unwrap_or_default(),
+            &parse_options,
+            io_client,
+            io_stats,
+        )
+        .await?;
+        return Table::empty(Some(Arc::new(Schema::try_from(&schema)?)));
+    }
+    let concated_table = tables_concat(collected_tables)?;
+    if let Some(limit) = limit
+        && concated_table.len() > limit
+    {
+        // Apply head in case that last chunk went over limit.
+        concated_table.head(limit)
+    } else {
+        Ok(concated_table)
+    }
+}
+
+/// Reads a single local CSV file in a streaming fashion.
+pub async fn stream_csv_local(
```

Comment:
This function doesn't look like it applies the global limit. Maybe we should move the logic from `read_csv_local` to here so it can be used for either reads or streams.